File: source/swift/src/Perms.swift
// Signed Permutation Cycle Counting

import Foundation

let VERSION = 4

// Generate initial permutation for each possible prefix
func enumPrefixes(_ n: Int, _ k: Int) -> [[Int]] {
    let rest = Array(0..<n)
    
    func pick(_ k: Int, _ prefix: [Int], _ rest: [Int]) -> [[Int]] {
        if k == 0 {
            return [prefix + rest]
        } else {
            var results: [[Int]] = []
            for x in rest {
                let sansX = rest.filter { $0 != x }
                results += pick(k - 1, prefix + [x], sansX)
            }
            return results
        }
    }
    
    return pick(k, [], rest)
}

// Distribute parcels across workers and combine results
func runParcels(_ n: Int, _ k: Int, _ cores: Int) -> [Int64] {
    let prefixes = enumPrefixes(n, k)
    let zero = Array(repeating: Int64(0), count: 2 * n)
    
    // Use parallelMap to process prefixes
    let results = parallelMap(cores, zero, { prefix in
        runPrefix(n, k, prefix)
    }, prefixes)
    
    // Combine results
    guard !results.isEmpty else { return zero }
    
    var combined = results[0]
    for i in 1..<results.count {
        for j in 0..<combined.count {
            combined[j] += results[i][j]
        }
    }
    
    return combined
}

// Entry point for cycle distribution computation
func runAll(_ n: Int, _ k: Int, _ cores: Int) -> [Int64] {
    return runParcels(n, k, cores)
}

// Parse command-line arguments
func parseArgs(_ args: [String]) -> (Int, Int, Int)? {
    guard args.count == 4 else {
        fputs("Error: Required arguments: n prefix cores\n", stderr)
        return nil
    }
    
    guard let n = Int(args[1]),
          let prefix = Int(args[2]),
          let cores = Int(args[3]) else {
        fputs("Error: Invalid arguments\n", stderr)
        return nil
    }
    
    return (n, prefix, cores)
}

// Entry point
@main
struct Main {
    static func main() {
        guard let args = parseArgs(CommandLine.arguments) else {
            exit(1)
        }
        
        let (n, k, cores) = args
        
        print("\(Relations.name) v\(VERSION), n = \(n), prefix = \(k), cores = \(cores)")
        
        let result = runAll(n, k, cores)
        checkResult(n, result)
    }
}

File: source/ocaml/src/perms.ml
(* Signed Permutation Cycle Counting *)

let version = 7

(* Generate initial permutation for each possible prefix *)
let enum_prefixes n k =
  let rest = List.init n Fun.id in
  let rec pick k prefix rest =
    if k = 0 then [ prefix @ rest ]
    else
      List.concat_map
        (fun x ->
          let sans_x = List.filter (( <> ) x) rest in
          pick (k - 1) (x :: prefix) sans_x)
        rest
  in
  pick k [] rest

(* Distribute work parcels and combine results *)
let run_parcels n k cores =
  let zero = Array.make (2 * n) 0 in
  enum_prefixes n k |> Parallel.map cores zero (Worker.run_prefix n k)
  |> function
  | [] -> zero
  | head :: tail -> List.fold_left (Array.map2 ( + )) head tail

(* Parse command-line arguments *)
let parse_args argv =
  match Array.to_list argv with
  | [ _; arg1; arg2; arg3 ] ->
    ( match List.map int_of_string_opt [ arg1; arg2; arg3 ] with
      | [ Some n; Some prefix; Some cores ] -> Some (n, prefix, cores)
      | _ -> None )
  | _ -> None

(* Main entry point *)
let () =
  match parse_args Sys.argv with
  | Some (n, prefix, cores) ->
      Printf.printf "%s v%d, n = %d, prefix = %d, cores = %d\n"
        Worker.name version n prefix cores;
      let result = run_parcels n prefix cores in
      Answers.check n result
  | None -> 
      Printf.eprintf "Required arguments: n prefix cores\n";
      exit 1
